src/c/modularizace/knihovny.md

```diff
@@ -1,57 +1,145 @@
+[zlepšit v programování](../co_dal.md#co-se-dále-naučit).
 
-Knihovny se obvykle nesdílí čistě jako archiv nebo adresář se zdrojovým kódem. Obvykle se setkáte s tím,
-že knihovna poskytuje dvě věci: 
+Aby programátoři mohli sdílet svůj kód s ostatními programátory, tak využívají tzv. **knihovny**
+(*libraries*). Knihovna je kód, který řeší nějakou ucelenou funkcionalitu (např.
+[vykreslování grafiky](https://www.libsdl.org/), [sazbu fontů](https://www.freetype.org/) nebo
+[kompresi dat](http://zlib.net/)) a obsahuje návod (dokumentaci), jak tento kód používat. Klíčové
+vlastnosti knihoven jsou znovupoužitelnost (můžeme je použít v různých programech) a abstrakce
+(nemusíme rozumět, jak knihovna funguje, pouze ji využijeme k vyřešení konkrétního problému).
 
-1. **Hlavičkové soubory** (*header files*) s příponou `.h`, které definují rozhraní, jak knihovnu používat.
-2. Soubory s příponami `.a` nebo `.so`, které obsahují již přeložené zdrojové soubory knihovny ve formě
-spustitelného kódu.
+> Knihovna není program – neobsahuje žádnou funkci `main` a nelze ji ani přímo spustit. V kontextu
+> jazyka *C* je knihovna typicky sada funkcí, struktur a globálních proměnných.
 
-> Více o knihovnách se můžete dozvědět například [zde](https://www.itnetwork.cz/cecko/linux/cecko-a-linux-staticke-a-dynamicke-knihovny).
+Například pokud bychom programovali hru, můžeme využít knihovny na vykreslení grafiky, na přehrávání
+zvuku, na snímání vstupu z klávesnice/myši atd. Náš kód se pak může zabývat zejména logikou hry a
+nemusí tolik řešit problémy, které již vyřešila spousta programátorů před námi.
 
-#### Použití knihoven pomocí `gcc`
-Abyste ve vašem programu použili nějakou knihovnu, musíte ji k vašemu programu tzv. **přilinkovat**.
-O to se stará tzv. **linker**, který za vás umí spustit překladač `gcc`.
+Na internetu můžete naleznout [tisice různých knihoven](https://github.com/kozross/awesome-c),
+které řeší rozlišné problémy.
 
-Dejme tomu, že chcete použít knihovnu s názvem `foo`, která obsahuje hlavičkové soubory v adresáři
-`/usr/foo/include` and zkompilovaný knihovní soubor v adresáři `/usr/foo/lib/libfoo.so`. Překladači
-`gcc` musíte říct, kde jsou umístěny knihovní soubory pomocí přepínače `-L`, které konkrétní soubory chcete
-přilinkovat pomocí přepínače `-l` a kde jsou umístěny hlavičkové soubory pomocí přepínače `-I`:
+## Sdílení knihoven
+Teoreticky bychom mohli knihovny používat prostě tak, že si nějakou najdeme na internetu, stáhneme
+její hlavičkové a zdrojové soubory k našeho programu a začneme je využívat. I když i tak to lze někdy
+udělat, není to obvyklé, protože tento přístup má spoustu nevýhod:
+- Jelikož obvykle nebudeme autory knihovny, kterou chceme použít, tak nemusíme ani být schopní
+danou knihovnu přeložit. Potřebuje daná knihovna konkrétní překladač nebo jeho specifické nastavení?
+Má závislosti na dalších knihovnách? Přeložit "cizí" knihovnu ze zdrojových souborů nemusí být
+zdaleka přímočaré.
```

Comment:
Tady bych dal jako domaci ukol prelozit OpenCV treba na Windows, tam je to obzvlast dobry opruz.